Gemfile

```diff
@@ -28,31 +46,64 @@ group :system_tests do
   gem 'beaker-puppet_install_helper', :require => false
 end
 
-is_x64 = Gem::Platform.local.cpu == 'x64'
-if is_x64
-  platform(:x64_mingw) do
-    gem "win32-dir", "~> 0.4.9", :require => false
-    gem "win32-process", "~> 0.7.4", :require => false
-    gem "win32-service", "~> 0.8.4", :require => false
-    gem "minitar", "~> 0.5.4", :require => false
-  end
-else
-  platform(:mingw) do
-    gem "win32-process", "~> 0.6.5", :require => false
-    gem "win32-service", "~> 0.7.2", :require => false
-    gem "minitar", "~> 0.5.4", :require => false
-    if RUBY_VERSION =~ /^1\./ then
-      gem "win32console", :require => false
+# The recommendation is for PROJECT_GEM_VERSION,
+# there are older ways of referencing these. Add
+# them all for compatibility reasons. Remove
+# later when no issues are known
+puppetversion = ENV['PUPPET_GEM_VERSION'] || ENV['GEM_PUPPET_VERSION'] || ENV['PUPPET_LOCATION'] || '>= 0'
+gem 'puppet', *location_for(puppetversion)
+puppet_gem_location = gem_type(puppetversion)
+
+if puppet_gem_location != :gem
+  gem "facter", *location_for(ENV['FACTER_GEM_VERSION'] || ENV['GEM_FACTER_VERSION'] || ENV['FACTER_LOCATION'] || '>= 0')
+  gem "hiera", *location_for(ENV['HIERA_GEM_VERSION'] || ENV['GEM_HIERA_VERSION'] || ENV['HIERA_LOCATION'] || '>= 0')
+end
+
+# For Windows dependencies, these could be required
+# based on the version of Puppet you are requiring.
+# Anything greater than v3.5.0 is going to have
+# Windows-specific dependencies dictated by the gem
+# itself.
+# The other scenario is when you are faking out the
+# Puppet Gem to use a local file path / git path.
+explicitly_require_windows_gems = false
+# This is not a perfect answer to the version check
+if puppet_gem_location != :gem || puppetversion < '3.5.0'
+  is_x64 = Gem::Platform.local.cpu == 'x64'
+  if is_x64
+    platform(:x64_mingw) do
+      explicitly_require_windows_gems = true
+    end
+  else
+    platform (:mingw) do
+      explicitly_require_windows_gems = true
     end
   end
 end
 
-ENV['GEM_PUPPET_VERSION'] ||= ENV['PUPPET_GEM_VERSION']
-puppetversion = ENV['GEM_PUPPET_VERSION']
-if puppetversion
-  gem 'puppet', *location_for(puppetversion)
-else
-  gem 'puppet', :require => false
+if explicitly_require_windows_gems
+  gem "ffi", "~> 1.9.0", :require => false
+  gem "win32-dir", "~> 0.3", :require => false
+  gem "win32-eventlog", "~> 0.5", :require => false
+  gem "win32-process", "~> 0.6", :require => false
+  gem "win32-security", "~> 0.1", :require => false
+  gem "win32-service", "~> 0.7", :require => false
+  gem "minitar", "~> 0.5.4", :require => false
+
+  if RUBY_VERSION =~ /^1\./ then
+    gem "win32console", :require => false
+  end
+
+  # Puppet less than 3.7.0 requires these
+  # Granted that anything over 3.5.0 won't make it in
+  # here anyway, but at least the check is accurate.
+  if puppet_gem_location != :gem || puppetversion < '3.7.0'
```

Comment:
This override needs to also apply to source, which likely means we don't need it.